design/nodes-machines-and-hosts.md

```diff
@@ -0,0 +1,213 @@
+MetalKube project was launched, and places the information in context
+as an aid to future contributors.
+
+### Goals
+
+- Explain Node, Machine, and BareMetalHost and how they relate to
+  MetalKube.
+- Encourage input on that design, to refine it.
+
+### Non-Goals
+
+- Discussion of user interface designs related to the resources.
+
+## Proposal
+
+The MetalKube project is building a hardware provisioning system with
+a Kubernetes-native API. It includes a provider implementation that
+fits into the API defined by the cluster-api SIG. It differs from
+other similar implementations by being self-hosted. Where the AWS or
+OpenStack providers run within Kubernetes, they allocate resources
+from a pool *outside* of the Kubernetes cluster. MetalKube will use
+Kubernetes for tracking available resources.
+
+### Implementation Details/Notes/Constraints
+
+#### Data Model
+
+`Node` objects represent a running instance of kubelet. Their status
+fields include basic information about the health of the environment
+in which kubelet is running. A user manipulates the Node to change the
+cluster, such as draining the node for maintenance. The definition of
+Node is owned by the Kubernetes core, and offers few opportunities to
+attach custom data.
+
+`Machine` objects represent a *request* for an instance of kubelet. A
+parallel design is the physical volume claim and physical volume in
+the storage APIs. The Machine CRD is owned by the cluster-api
+SIG. Machine objects are managed by the cluster-api controller using
+an *actuator* which works as a plugin driver for the
+controller. Machine objects include a "provider spec" field to allow
+the actuator to store custom data. The lifecycle of a Machine is based
+on the desired size of the cluster. That is, a Machine object does not
+typically exist in the database when it does not represent a request
+to expand the cluster (see caveats below in the
+[Alternatives](#alternatives-optional) section.
+
+`BareMetalHost` objects represents a physical computer, including its
+hardware inventory and information to access the onboard management
+controller. The BareMetalHost (or "host") CRD is defined in the
+[baremetal-operator](https://github.com/metalkube/baremetal-operator/)
+from the MetalKube project. The CRD is made up of completely custom
+data, needed to fully realize a Kubernetes-native API for a
+self-hosted baremetal provisioning tool. The lifecycle of a host is
+based on the availability of hardware, rather than its use as part of
+the cluster.
+
+#### Growing the Cluster
+
+The admin user interacts with MetalKube by first registering hosts to
+be part of the inventory available to the provisioning system. Each
+host is then matched against known hardware profiles (comparing
+information such as NICs, storage, CPUs, RAM, etc.). Hosts matching
+known profiles are marked as being ready to be used. The admin (or
+autoscaler) can then create a new Machine object as a request to grow
+the cluster.
```

Comment:
Or increments the count in a MachineSet, but maybe that detail isn't really important here ...

Reply:
That's a bit of a detail, but I went ahead and included it.